### PR TITLE
[Requested] Serenity changed to proper medical exosuit.

### PR DESCRIPTION
**Pull request by Mechoid**
Will be discussed at staff meeting, PR made by request.

Serenity tweaked to be a medical subtype, weapons cannot fit in universal slots.

Serenity weapon slot converted to a hull slot.

Exosuit punching converted to general ability, if melee_can_hit is TRUE by default.

### CODE REVIEW

**Comment by Sypsoti** [comment]
General approval at the 2/19 staff meeting.

File: code/game/mecha/mecha.dm [hunk 1]
```diff
@@ -674,7 +678,92 @@
 		if(src_object in view(2, src))
 			return STATUS_UPDATE //if they're close enough, allow the occupant to see the screen through the viewport or whatever.
 
-/obj/mecha/proc/melee_action(atom/target)
+/obj/mecha/proc/melee_action(atom/T)
+	if(internal_damage&MECHA_INT_CONTROL_LOST)
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:682]: Spacing

File: code/game/mecha/mecha.dm [hunk 2]
```diff
@@ -674,7 +678,92 @@
...
 			return STATUS_UPDATE //if they're close enough, allow the occupant to see the screen through the viewport or whatever.
 
-/obj/mecha/proc/melee_action(atom/target)
+/obj/mecha/proc/melee_action(atom/T)
+	if(internal_damage&MECHA_INT_CONTROL_LOST)
+		T = safepick(oview(1,src))
+	if(!melee_can_hit)
+		return
+
+	if(istype(T, /mob/living))
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:687]: isliving

File: code/game/mecha/mecha.dm [hunk 3]
```diff
@@ -674,7 +678,92 @@
...
+		var/mob/living/M = T
+		if(src.occupant.a_intent == I_HURT || istype(src.occupant, /mob/living/carbon/brain)) //Brains cannot change intents; Exo-piloting brains lack any form of physical feedback for control, limiting the ability to 'play nice'.
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:689]: isbrain

File: code/game/mecha/mecha.dm [hunk 4]
```diff
@@ -674,7 +678,92 @@
...
+			playsound(src, 'sound/weapons/heavysmash.ogg', 50, 1)
+			do_attack_animation(T)
+			if(damtype == "brute")
+				step_away(M,src,15)
+
+			if(ishuman(T))
+				var/mob/living/carbon/human/H = T
+
+				var/obj/item/organ/external/temp = H.get_organ(pick(BP_TORSO, BP_TORSO, BP_TORSO, BP_HEAD))
+				if(temp)
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:699]:
```suggestion
				if(!istype(temp))
					return
				// dedent organ-smashing
```

File: code/game/mecha/mecha.dm [hunk 5]
```diff
@@ -674,7 +678,92 @@
...
+					var/update = 0
+					switch(damtype)
+						if("brute")
+							H.Paralyse(1)
+							update |= temp.take_damage(rand(force/2, force), 0)
+						if("fire")
+							update |= temp.take_damage(0, rand(force/2, force))
+						if("tox")
+							if(H.reagents)
+								if(H.reagents.get_reagent_amount("carpotoxin") + force < force*2)
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:709]: `[...] + force < force*2` is adding `force` to both sides

File: code/game/mecha/mecha.dm [hunk 6]
```diff
@@ -674,7 +678,92 @@
...
+									H.reagents.add_reagent("cryptobiolin", force)
+						if("halloss")
+							H.stun_effect_act(1, force / 2, BP_TORSO, src)
+						else
+							return
+					if(update)	H.UpdateDamageIcon()
+				H.updatehealth()
+
+			else
+				switch(damtype)
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:721]: This pains me to read but the only way I can conceive of fixing it is barely any more readable

File: code/game/mecha/mecha.dm [hunk 7]
```diff
@@ -674,7 +678,92 @@
...
+			src.visible_message("[src] pushes [T] out of the way.")
+
+		melee_can_hit = 0
+		if(do_after(melee_cooldown))
+			melee_can_hit = 1
+		return
+
+	else
+		if(istype(T, /obj/machinery/disposal)) // Stops mechs from climbing into disposals
+			return
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:750]: Early return

File: code/game/mecha/mecha.dm [hunk 8]
```diff
@@ -674,7 +678,92 @@
...
+		if(src.occupant.a_intent == I_HURT || istype(src.occupant, /mob/living/carbon/brain)) // Don't smash unless we mean it
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:751]: Flip this to an early return and dedent

File: code/game/mecha/mecha.dm [hunk 9]
```diff
@@ -674,7 +678,92 @@
...
+
+				if(istype(T, /obj/structure/girder))
+					T:take_damage(force * 3) //Girders have 200 health by default. Steel, non-reinforced walls take four combat exosuit punches, girders take (with this value-mod) two, girders took five without.
+				else
+					T:take_damage(force)
+
+				melee_can_hit = 0
+
+				if(do_after(melee_cooldown))
+					melee_can_hit = 1
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:766]: This code concludes both sides of the isliving()

File: code/game/mecha/mecha.dm [hunk 10]
```diff
@@ -674,7 +678,92 @@
...
+			melee_can_hit = 1
+		return
+
+	else
+		if(istype(T, /obj/machinery/disposal)) // Stops mechs from climbing into disposals
+			return
+		if(src.occupant.a_intent == I_HURT || istype(src.occupant, /mob/living/carbon/brain)) // Don't smash unless we mean it
+			if(damtype == "brute")
+				src.occupant_message("You hit [T].")
+				src.visible_message("<font color='red'><b>[src.name] hits [T]</b></font>")
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:754]: SPAN_DANGER()?

File: code/game/mecha/mecha.dm [hunk 11]
```diff
@@ -674,7 +678,92 @@
...
+						if(M.reagents)
+							if(M.reagents.get_reagent_amount("carpotoxin") + force < force*2)
+								M.reagents.add_reagent("carpotoxin", force)
+							if(M.reagents.get_reagent_amount("cryptobiolin") + force < force*2)
+								M.reagents.add_reagent("cryptobiolin", force)
+					else
+						return
+				M.updatehealth()
+			src.occupant_message("You hit [T].")
+			src.visible_message("<font color='red'><b>[src.name] hits [T].</b></font>")
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:737]: SPAN_DANGER()?

File: code/game/mecha/medical/odysseus.dm [hunk 12]
```diff
@@ -27,11 +27,12 @@
...
-			H.recalculate_vis()
+		if(istype(H))
+			if(H.glasses)
+				occupant_message("<font color='red'>[H.glasses] prevent you from using [src] [hud]</font>")
+			else
+				H.glasses = hud
+				H.recalculate_vis()
 		return 1
 	else
 		return 0
```
Comment (Atermonera) [inline code/game/mecha/medical/odysseus.dm:38]:
```suggestion
	. = ..()
	if(istype(H))
		if(H.glasses)
			occupant_message("<font color='red'>[H.glasses] prevent you from using [src] [hud]</font>")
		else
			H.glasses = hud
			H.recalculate_vis()

File: code/game/mecha/medical/serenity.dm [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+	var/obj/item/clothing/glasses/hud/health/mech/hud
+
+/obj/mecha/medical/serenity/Initialize()
+	. = ..()
+	hud = new /obj/item/clothing/glasses/hud/health/mech(src)
+	return
+
+/obj/mecha/medical/serenity/moved_inside(var/mob/living/carbon/human/H as mob)
+	if(..())
```
Comment (Atermonera) [inline code/game/mecha/medical/serenity.dm:37]: See my other suggestion about move_inside - actually hey wait just put this on `/obj/mecha/medical`, the code is duplicated between both medical mecha (Should the hud go on mecha/medical too?)

File: code/game/mecha/medical/serenity.dm [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+	else
+		return 0
+
+/obj/mecha/medical/serenity/go_out()
+	if(ishuman(occupant))
+		var/mob/living/carbon/human/H = occupant
+		if(H.glasses == hud)
+			H.glasses = null
+			H.recalculate_vis()
+	..()
```
Comment (Atermonera) [inline code/game/mecha/medical/serenity.dm:54]: `return ..()`

File: code/game/mecha/medical/serenity.dm [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+	return
+
+/obj/mecha/medical/serenity/melee_action(atom/T)
```
Comment (Atermonera) [inline code/game/mecha/medical/serenity.dm:57]: How much of this proc is copy pasted?

**Review by Atermonera — CHANGES_REQUESTED** [review]
I want to state for the record that I am being this nitpicky not out of malice towards you, but towards the old code you're bringing before me.

File: code/game/mecha/mecha.dm [hunk 16]
```diff
@@ -696,7 +696,7 @@
 				var/mob/living/carbon/human/H = T
 
 				var/obj/item/organ/external/temp = H.get_organ(pick(BP_TORSO, BP_TORSO, BP_TORSO, BP_HEAD))
-				if(temp)
+				if(!istype(temp))
```
Comment (Atermonera) [inline code/game/mecha/mecha.dm:699]: Temp... isn't an organ?

File: code/game/mecha/mecha.dm (hunk 7 above)
Comment (Atermonera) [reply code/game/mecha/mecha.dm:745]: (Move this to the top of the proc so its more distinct)

File: code/game/mecha/medical/medical.dm [hunk 17]
```diff
@@ -17,12 +18,35 @@
...
+			H.recalculate_vis()
+
+/obj/mecha/medical/odysseus/go_out()
+	if(ishuman(occupant))
+		var/mob/living/carbon/human/H = occupant
+		if(H.glasses == hud)
+			H.glasses = null
+			H.recalculate_vis()
+	..()
+	return
```
Comment (Atermonera) [inline code/game/mecha/medical/medical.dm:48]: `. = ..()`?

**Review by Atermonera — COMMENTED** [review]
The temp thing might be a regression but the others are just nits, this is pretty close to being good.

**Review by Atermonera — APPROVED** [review]

**Pull request merged**